e2e/cephfs.go

```diff
@@ -172,549 +194,541 @@ var _ = Describe("cephfs", func() {
+				_, stdErr := execCommandInPodAndAllowFail(f, fmt.Sprintf("echo 'Hello World' > %s", filePath), app.Namespace, &opt)
+				readOnlyErr := fmt.Sprintf("cannot create %s: Read-only file system", filePath)
+				if !strings.Contains(stdErr, readOnlyErr) {
+					e2elog.Failf(stdErr)
+				}
 
-				By("Mount pvc as readonly in pod", func() {
-					// create pvc and bind it to an app
+				// delete pvc and app
+				err = deletePVCAndApp("", f, pvc, app)
+				if err != nil {
+					e2elog.Failf("failed to delete pvc or application %v", err)
+				}
+			})
+
+			By("create a PVC clone and bind it to an app", func() {
+				v, err := f.ClientSet.Discovery().ServerVersion()
+				if err != nil {
+					e2elog.Failf("failed to get server version with error %v", err)
+				}
+				// snapshot beta is only supported from v1.17+
+				if v.Major > "1" || (v.Major == "1" && v.Minor >= "17") {
+					var wg sync.WaitGroup
+					totalCount := 3
+					// totalSubvolumes represents the subvolumes in backend
+					// always totalCount+parentPVC
+					totalSubvolumes := totalCount + 1
+					wg.Add(totalCount)
+					err = createCephFSSnapshotClass(f)
+					if err != nil {
+						e2elog.Failf("failed to delete cephfs storageclass %v", err)
+					}
 					pvc, err := loadPVC(pvcPath)
 					if err != nil {
-						Fail(err.Error())
+						e2elog.Failf("failed to load pvc %v", err)
 					}
 
 					pvc.Namespace = f.UniqueName
+					err = createPVCAndvalidatePV(f.ClientSet, pvc, deployTimeout)
+					if err != nil {
+						e2elog.Failf("failed to create pvc %v", err)
+					}
 
 					app, err := loadApp(appPath)
 					if err != nil {
-						Fail(err.Error())
+						e2elog.Failf("failed to load application %v", err)
 					}
 
 					app.Namespace = f.UniqueName
-					label := map[string]string{
-						"app": app.Name,
-					}
-					app.Labels = label
 					app.Spec.Volumes[0].PersistentVolumeClaim.ClaimName = pvc.Name
-					app.Spec.Volumes[0].PersistentVolumeClaim.ReadOnly = true
-					err = createPVCAndApp("", f, pvc, app, deployTimeout)
-					if err != nil {
-						Fail(err.Error())
-					}
-
-					opt := metav1.ListOptions{
-						LabelSelector: fmt.Sprintf("app=%s", app.Name),
+					wErr := writeDataInPod(app, f)
+					if wErr != nil {
+						e2elog.Failf("failed to  write data %v", wErr)
 					}
 
-					filePath := app.Spec.Containers[0].VolumeMounts[0].MountPath + "/test"
-					_, stdErr := execCommandInPodAndAllowFail(f, fmt.Sprintf("echo 'Hello World' > %s", filePath), app.Namespace, &opt)
-					readOnlyErr := fmt.Sprintf("cannot create %s: Read-only file system", filePath)
-					if !strings.Contains(stdErr, readOnlyErr) {
-						Fail(stdErr)
-					}
-
-					// delete pvc and app
-					err = deletePVCAndApp("", f, pvc, app)
+					snap := getSnapshot(snapshotPath)
+					snap.Namespace = f.UniqueName
+					snap.Spec.Source.PersistentVolumeClaimName = &pvc.Name
+					// create snapshot
+					for i := 0; i < totalCount; i++ {
```

Comment:
As this is a refactoring PR, isnt it a good idea to have this in a common function with arg `CephFS or RBD` ? so that we can reuse the same ?

Reply:
Also we keep different functions for cephfs and RBD resource creation. I believe its better to have a common function with argument controlled.

Reply:
we can do it, it can be a followup PR I don't want to do it in this PR. This PR mainly concentrates on reorganizing the function to there own files and improve the logging.